### ISSUE TITLE
Performance drop from 4.13.10 to 5.0.1

### ISSUE THREAD

**Issue by dackmin** [Automattic/mongoose#6065]
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

After upgrading from `4.13.10` to `5.0.1` on production environment, we noticed a real performance drop in all requests using mongoose models : ~200ms per response. We didn't release any code change apart from the `mongoose` dependency upgrade. And we don't use any particular feature of mongoose apart from models & schemas.
After applying a fix to revert to `4.13.10`, everything went back to normal.

Deploying with `5.0.1` : 

![capture d ecran 2018-01-29 a 07 46 52](https://user-images.githubusercontent.com/1873323/35498178-835f0818-04cd-11e8-95ed-bc28f43f93dd.png)

Reverting back to `4.13.10` : 

![capture d ecran 2018-01-29 a 08 12 06](https://user-images.githubusercontent.com/1873323/35498190-8c4776ae-04cd-11e8-8098-704a294e424a.png)

**If the current behavior is a bug, please provide the steps to reproduce.**

More of a performance issue than a bug, so I'm not sure I will be able to reproduce it outside of a  production environment. If someone knows any way to benchmark this I will be happy to try 😊 

**What is the expected behavior?**

Having response times around ~20ms for 150 RPS.

**Please mention your node.js, mongoose and MongoDB version.**

node.js 8.9.4, mongoose 5.0.1, MongoDB 3.4.10.
If it's of any use, we are on a cloud managed (Atlas) sharded cluster.

**Comment by Jeff-Lewis** [comment]
Interesting that both node and MongoDB times increased with 5.0.1. Are you able to isolate or narrow it down to a couple example requests that take significantly longer?

**Comment by vkarpov15** [comment]
@dackmin this is definitely not expected and a significant problem, but without any sample code this issue simply isn't actionable. Can you provide any examples of endpoints that are much slower in 5.0.1 than 4.13.x?

**Comment by vkarpov15** [comment]
Also, since you're on atlas, can you check your mongodb cloud manager and see if you see any evidence of index misses? One potential issue is that 5.x removed the `retainKeyOrder` option, which can impact compound indexes.

**Comment by dackmin** [comment]
@vkarpov15 that's the first thing I did when noticed the performance drop, everything is fine on atlas, no index is missing, no memory/cpu/iops/page faults or any particular increase during that time. I will try to run our cursed api version on App Engine for 10% of our traffic to see if I can isolate anything, since I cannot have more than 2 days of retention on new relic.

**Comment by dackmin** [comment]
As I thought, new relic is not a great ally here. The only difference between an instance with `4.13.10` and `5.0.1` is a globally higher response time.

Instance with `4.13.10` : 
![capture d ecran 2018-01-31 a 08 57 08](https://user-images.githubusercontent.com/1873323/35611268-c73e820a-0664-11e8-83b8-3da4aca1b7b9.png)

Instance with `5.0.1` : 
![capture d ecran 2018-01-31 a 08 51 26](https://user-images.githubusercontent.com/1873323/35611275-d22ea7b2-0664-11e8-90c5-8314e918d8c7.png)

The middleware stated as `anonymous` is actually using a compound index on a quite large collection (~150M documents) : 
```coffeescript
    # If user id exists, find user by id
    if user_id?
        promise = Person.findById user_id

    # If not, try to find it by ip & fingerprint
    else if fingerprint?
        promise = Person.findOne ip: ip, fingerprint: fingerprint
```

Both normal & compound index are heavily used : 
![capture d ecran 2018-01-31 a 09 03 22](https://user-images.githubusercontent.com/1873323/35611546-a160d37a-0665-11e8-980f-d37309db753a.png)
![capture d ecran 2018-01-31 a 09 03 28](https://user-images.githubusercontent.com/1873323/35611547-a17c738c-0665-11e8-9ac8-cef232ee11cf.png)

**Comment by vkarpov15** [comment]
Can you isolate any one endpoint that is much slower with 5.0 versus 4.13? That would be a good place to start figuring out where this issue is.

**Comment by hilkeheremans** [comment]
I am also experiencing this issue, for some _very simple_ use cases on a (for now) MongoDB Atlas sandbox environment. Queries on Mongoose 5.x appear to take between 60-120% longer than on Mongoose 4.x.

One of the collections has a compound index, but the other collection suffering from the exact same problem only has a simple index on a single field.

Guessing this might be Atlas and/or ReplicaSet related.

**Comment by mrkmrtns** [comment]
I have similar problem, altough I'm using latest mongodb native driver, not mongoose. With native driver version 3.* simple findOne takes 40ms (server have no load) with driver 2.* it takes 5ms. As Mongoose 4 uses mongodb 2.* as dependancy and 5 uses 3.* then we may have the same route cause.

In addition I have two different projects, one uses ReplicaSet and another standalone server. Affected seems only be ReplicaSet version. With standalone perf drop was only few % with upgrading the latest driver. 

Hope this helps to get root and fix because as I reverted mongodb driver for ReplicaSet to 2.* because of it.

**Comment by dackmin** [comment]
@vkarpov15 that is what is in the screenshots 😄 see `post /api/v2/track`

The endpoint shown also has a simple `.findOne` but is using a different collection. Reading previous comments, it might have something to do with replica sets & driver 3.x. 

I will be glad to open an issue on mongodb driver repository, but you guys might have a louder voice than mine there 😊

**Comment by lustoykov** [comment]
I can confirm incredibly slow performance for mongodb 3.6.2 and mongoose 5.0.4, SSL is on, virtually no entries in the database, a simple request takes about 200-500ms for EC2 to EC2 instance. I'm unfortunately not able to isolate the issue more than that. Previously mongoose 4.X worked just fine.

**Comment by mrkmrtns** [comment]
One more comment from my side, first server (standalone server, which has no issues with new driver) is not using SSL. Secondone is using SSL (replicaset, which has issues with new driver). So it can be in my case also SSL related or replicaset related or both.

**Comment by vkarpov15** [comment]
@lustoykov is this performance regression sustained or does it just happen on the first couple requests?

**Comment by lustoykov** [comment]
@vkarpov15 it is sustained

**Comment by hilkeheremans** [comment]
I'm willing to try and reproduce these issues in a public repo this week, which should sufficiently establish the issue & help everyone trace the problem more precisely.

Proposed test cases:
- Drivers: raw mongodb driver, `2.x` vs `3.x`, Mongoose `4.x` vs `5.x`
- Database configurations:
   - MongoDB single instance, local (Docker)
   - MongoDB single instance, remote server (Docker)
   - MongoDB Replicaset, no SSL, local (Docker)
   - MongoDB Replicaset, no SSL, remote server (Docker)
   - MongoDB Replicaset, SSL, local (Docker)
   - MongoDB Replicaset, SSL, remote (Docker)
   - MongoDB Replicaset, SSL, remote (Atlas)
- Queries: tbd, it seems to occur on most queries. Depends on how quickly we can repro the problem

Anything I might miss here?

**Comment by mrkmrtns** [comment]
@hilkeheremans maybe split single instance also by no SSL and SSL case

**Comment by vkarpov15** [comment]
@hilkeheremans I'd focus on mongodb single instance, remote server with Atlas, and answering the question of which queries are slower. Single instance vs replica set, SSL vs not, local vs remote, I have a feeling that those distinctions are beside the point here.

**Comment by simllll** [comment]
I experience similar issues on my production environment since yesterday (mongodb with 2 shards, mongos running directly on the webservers). The only change that has been deployed to the webservers is that ulimit got increased (ulimit -HSn 20000) for mongos. I'm not sure if this is related, I will try to find some time to debug this today.

**Comment by phgroe** [comment]
Performance degradation confirmed on one of our projects here where I tried to upgrade from 4.13.7 to 5.0.6, see above reference. The database test suite for all the models and DAOs takes more than twice the time to complete with a v5 Mongoose compared to before. This is running on a local v3.4.10 MongoDB.

**Comment by simllll** [comment]
Can I help to track this down? I get a lot of "connection 19 to 127.0.0.1:27017 timed out" since then, and even though it got much better with the config changes I added to my last comment, it's still not ideal.
Are there any hints what could cause this? Or is it better to downgrade to 4.x for now?

**Comment by lustoykov** [comment]
I actually have a local repository with mongoose 5.0.6 that queries a brand new mongodb 3.6.2 with mongoose > 5.0.6, the server was specifically setup to reproduce the issue, but I'm not sure if the database is slow now or the roundtrip time from Munich (Home) to California (EC2) is too significant.

@vkarpov15 if you think that I can help, I can invite you as a collaborator and give you password & IP to the test DB, it would take you about 2 mins to run the whole setup. But I have a feeling @phgroe's reference is more valuable here.

**Comment by vkarpov15** [comment]
@lustoykov yes please add me as a collaborator. I don't need the password and IP for the test DB, I'd imagine I can just point it at a MongoDB Atlas instance and see the same performance degradation.

**Comment by hilkeheremans** [comment]
As mentioned above, the problem occurs just as well in native mongodb `3.x`, and does not seem to occur on a single mongodb instance.

Interestingly, it also only occurs when querying on indexes fields -- non-indexed queries causing full document scans show no performance degradation vs `2.x` (but are, of course, very slow by themselves).

**Comment by larsdecker** [comment]
We have the same performance drop on our production environment. The issue occurs also with mongoose 5.0.3, but only when it run for a while. When the issue occurs, i see that the CPU usage it a little bit higher and the memory consumption is about 100 MB higher than normal. 

We use a ReplicaSet with 3 Servers. Sometimes the queries are very fast and some queries we have a huge performance drop by 3-6 Seconds. The issue occurs on every API-Route, that interact with the MongoDB.

We use a ReplicaSet with 3 Servers for the production environment. On my local Setup a can not measure any difference.

**Comment by misshula** [comment]
We too are noticing this on our replica set on Atlas. 

We do have a simple script that can easily recreate the issue. Essentially it's: 
1. create documents
2. get a cursor and loop over docs
3. for each doc, just query `findOne` on that document
4. measure time for each query. 

you'll notice mongoose@5 is about 4x slower than mongoose@4  (I assume it's the driver from 2.x -> 3.x - but I may be wrong) 

originally thought it was a connection string issue - but not combinations of connection options have solved the issue. (other than the obvious `autoIndex: false` that just speeds up boot time)  

@vkarpov15 let me know if you would like my script.

**Comment by vkarpov15** [comment]
@misshula absolutely, your script would save me a lot of time

@lustoykov I got your repo but it isn't particularly helpful, doesn't actually include any timing measurements or comparisons, just runs some queries every 30 seconds and prints the results.

**Comment by vkarpov15** [comment]
@misshula I tried piecing together your script because I'm getting frustrated with the lack of actionable information in this thread. Across the runs I've done mongoose 5 seems to be slightly faster than mongoose 4. Below is the script:

```javascript
const mongoose4 = require('mongoose');
// Via symlink
const mongoose5 = require('mongoose5');

console.log(mongoose4.version);
console.log(mongoose5.version);

run().catch(error => console.error(error.stack));

async function run() {
  mongoose4.Promise = global.Promise;
  await mongoose4.connect('mongodb://localhost:27017/m4', { useMongoClient: true });
  await mongoose5.connect('mongodb://localhost:27017/m5');

  const Model4 = mongoose4.model('Test', new mongoose4.Schema({
    name: String
  }));
  const Model5 = mongoose5.model('Test', new mongoose5.Schema({
    name: String
  }));

  const reps = 100;

  let m5total = 0;
  for (let i = 0; i < reps; ++i) {
    m5total += await profile(Model5);
  }

  console.log('m5', m5total, 'average', m5total / reps);

  let m4total = 0;
  for (let i = 0; i < reps; ++i) {
    m4total += await profile(Model4);
  }

  console.log('m4', m4total, 'average', m4total / reps);

  console.log('mongoose 5 perf reduction in ms:', (m5total - m4total) / reps);
}

async function profile(Model) { 
  await Model.deleteMany({});
  for (let i = 0; i < 500; ++i) {
    await Model.create({ name: 'test' });
  }
  
  let totalMS = 0;

  const cursor = Model.find().cursor();
  for (let doc = await cursor.next(); doc != null; doc = await cursor.next()) {
    const start = Date.now();
    await Model.findById(doc._id);
    totalMS += Date.now() - start;
  }

  return totalMS;
}
```

Sample output:

```
$ node gh-6065.js 
4.13.11
5.0.8-pre
m5 7223 average 72.23
m4 7941 average 79.41
mongoose 5 perf reduction in ms: -7.18
^C
$ 
```

Perhaps I'm missing something, can you provide your script or modify mine to show a performance degradation in mongoose 5?

Easiest way I've found to have mongoose 4 and mongoose 5 simultaneously is to `npm install mongoose@4.x` and then `git clone` the mongoose repo and symlink it into your `node_modules` with `ln -s ~/path/to/mongoose ./node_modules/mongoose5`

**Comment by misshula** [comment]
@vkarpov15  

here is a script that I am using. 

Prereqs:
1. this needs to be on a Replica Set on Atlas (as mentioned above)
2. you will need to insert some sample docs (I've called my model `Thing` below)

```javascript
const moment = require('moment');
const mongoose = require('mongoose');
const mongodb = require('mongodb');
const Thing = require('./thing.model');

let client;

let startTime;
let times = 0;

function timePassed() {
    const diff = moment().diff(startTime);
    return [{ diff, times: ++times, adv: diff / times }];
}

const benchmark = (Model) => {

    const cursor = Model.find({}).sort({ _id: 1 }).cursor();

    return cursor.eachAsync(function(document) {
        console.log(document._id, timePassed());

        // try one of these below (it doesn't actually matter)

        // 1. this is trying mongoose connect
        return Thing.find({ _id: document._id });

        // 2. this is trying mongodb directly
        // return client.db('dbname').collection('thingss').findOne({ _id: document._id });
    });
};

Promise.all([
    // the mongouri needs to be a Replica Set on Atlas
    mongoose.connect(<mongouri>, <options>),
    mongodb.connect(<mongouri>, <options>)
])
.then(([mongooseClient,_client]) => {
    client = _client;
    startTime = new Date();
    return benchmark(Thing);
})
.then(() => {
    console.log(timePassed());
    process.exit(0);
})
.catch((err) => {
    console.error('An error occurred');
    console.error(err);
    process.exit(-1);
});

```

output with `mongoose@5`:
```
5a85d60a686806000526d758 [ { diff: 136, times: 1, adv: 136 } ]
5a85e668686806000526d759 [ { diff: 248, times: 2, adv: 124 } ]
5a85e727686806000526d75a [ { diff: 354, times: 3, adv: 118 } ]
5a86ef928b7d800005a5bf7d [ { diff: 460, times: 4, adv: 115 } ]
5a86f69b8b7d800005a5bf7e [ { diff: 565, times: 5, adv: 113 } ]
5a87215d8b7d800005a5bf7f [ { diff: 671, times: 6, adv: 111.83333333333333 } ]
5a89059f8b7d800005a5bf80 [ { diff: 780, times: 7, adv: 111.42857142857143 } ]
5a89078f8b7d800005a5bf81 [ { diff: 892, times: 8, adv: 111.5 } ]
5a890a528b7d800005a5bf82 [ { diff: 997, times: 9, adv: 110.77777777777777 } ]
5a890a928b7d800005a5bf83 [ { diff: 1101, times: 10, adv: 110.1 } ]
5a890af28b7d800005a5bf84 [ { diff: 1207, times: 11, adv: 109.72727272727273 } ]
5a890b888b7d800005a5bf85 [ { diff: 1313, times: 12, adv: 109.41666666666667 } ]
5a8d8cd28bff280005d8b32b [ { diff: 1419, times: 13, adv: 109.15384615384616 } ]
5a8d93128bff280005d8b32c [ { diff: 1526, times: 14, adv: 109 } ]
5a9020dffc42840005ed247c [ { diff: 1631, times: 15, adv: 108.73333333333333 } ]
5a902116fc42840005ed247d [ { diff: 1737, times: 16, adv: 108.5625 } ]
5a9032e2fc42840005ed247e [ { diff: 1843, times: 17, adv: 108.41176470588235 } ]
5a94337bdd61c60005df50a4 [ { diff: 1948, times: 18, adv: 108.22222222222223 } ]
[ { diff: 2054, times: 19, adv: 108.10526315789474 } ]
```

output with `mongoose@4`:
```
5a85d60a686806000526d758 [ { diff: 54, times: 1, adv: 54 } ]
5a85e668686806000526d759 [ { diff: 96, times: 2, adv: 48 } ]
5a85e727686806000526d75a [ { diff: 132, times: 3, adv: 44 } ]
5a86ef928b7d800005a5bf7d [ { diff: 168, times: 4, adv: 42 } ]
5a86f69b8b7d800005a5bf7e [ { diff: 203, times: 5, adv: 40.6 } ]
5a87215d8b7d800005a5bf7f [ { diff: 241, times: 6, adv: 40.166666666666664 } ]
5a89059f8b7d800005a5bf80 [ { diff: 279, times: 7, adv: 39.857142857142854 } ]
5a89078f8b7d800005a5bf81 [ { diff: 318, times: 8, adv: 39.75 } ]
5a890a528b7d800005a5bf82 [ { diff: 353, times: 9, adv: 39.22222222222222 } ]
5a890a928b7d800005a5bf83 [ { diff: 387, times: 10, adv: 38.7 } ]
5a890af28b7d800005a5bf84 [ { diff: 421, times: 11, adv: 38.27272727272727 } ]
5a890b888b7d800005a5bf85 [ { diff: 455, times: 12, adv: 37.916666666666664 } ]
5a8d8cd28bff280005d8b32b [ { diff: 490, times: 13, adv: 37.69230769230769 } ]
5a8d93128bff280005d8b32c [ { diff: 524, times: 14, adv: 37.42857142857143 } ]
5a9020dffc42840005ed247c [ { diff: 558, times: 15, adv: 37.2 } ]
5a902116fc42840005ed247d [ { diff: 593, times: 16, adv: 37.0625 } ]
5a9032e2fc42840005ed247e [ { diff: 629, times: 17, adv: 37 } ]
5a94337bdd61c60005df50a4 [ { diff: 663, times: 18, adv: 36.833333333333336 } ]
[ { diff: 698, times: 19, adv: 36.73684210526316 } ]
```

**Comment by misshula** [comment]
@vkarpov15 your script looks good - just need to connect to an Atlas db....

**Comment by simllll** [comment]
It's not only atlas db, in my case It's a self hosted 2 shards (each 1 replica) cluster. If I run the script I get:
Running on localhost through tunnel: 
```
4.13.11
5.0.7
m5 150269 average 15026.9
m4 147987 average 14798.7
mongoose 5 perf reduction in ms: 228.2
```

Running on a webserver: 
```
4.13.11
5.0.7
m5 8138 average 813.8
m4 7700 average 770
mongoose 5 perf reduction in ms: 43.8
```

I wonder why are my average times are so much higher than yours.

regards
Simon

**Comment by vkarpov15** [comment]
@simllll I haven't tried with mongos, but I have been able to reproduce a significant performance degradation in the mongodb node driver when working with Atlas. See https://jira.mongodb.org/browse/NODE-1357 .

**Comment by larsdecker** [comment]
Hopefully, one of MongoDB will fix the bug quickly.

**Comment by mrkmrtns** [comment]
Bug is fixed by applying this patch:
https://github.com/mongodb-js/mongodb-core/commit/e5ff9275c84b0ba8d35ea8c937586eccc90559df

Release coming soon.
https://jira.mongodb.org/browse/NODE-1357

**Comment by vkarpov15** [comment]
@mbroadst is a champ and fixed the issue over the weekend as @mrkmrtns  pointed out. we'll have a new mongoose release as soon as the mongodb driver pushes a release with this fix :tada:

**Comment by mbroadst** [comment]
👋  Sorry for the trouble y'all! Somewhat unexpected behavior from node there, hope it didn't cause too much trouble. `mongodb` and `mongodb-core` v3.0.4 have been released, and are ready to go.

**Comment by dackmin** [comment]
will try with mongoose the second it rolls out 🥇

**Comment by vkarpov15** [comment]
Shipped 5.0.9 with this fix :+1:

**Comment by phgroe** [comment]
Confirmed here, with 5.0.9 our tests aren't slower than with 4.13.7, in fact they are even a little bit faster :)

Well done and thanks a lot!
